src/future/batch_semaphore.rs

```diff
@@ -28,11 +30,178 @@ impl Waiter {
+            permit_clocks.push_back((num_permits, VectorClock::new()));
+        }
+        Self::Permits {
+            num_available: num_permits,
+            permit_clocks,
+            last_acquire: VectorClock::new(),
+        }
+    }
+
+    const fn const_new(num_permits: usize) -> Self {
+        if num_permits == usize::MAX {
+            return Self::Infinity(VectorClock::new());
+        }
+        // TODO: The VecDeque cannot be populated in a const fn, due to
+        // TODO: allocation. If this is ever needed, another variant can be
+        // TODO: added to `PermitsAvailable`, which would express that a
+        // TODO: certain number of permits are available with a zero clock.
+        assert!(num_permits == 0);
+        Self::Permits {
+            num_available: num_permits,
+            permit_clocks: VecDeque::new(),
+            last_acquire: VectorClock::new(),
+        }
+    }
+
+    fn available(&self) -> usize {
+        match self {
+            Self::Infinity(_) => usize::MAX,
+            Self::Permits { num_available, .. } => *num_available,
+        }
+    }
+
+    fn failure_clock(&self) -> Option<&VectorClock> {
+        match self {
+            Self::Infinity(_) => None,
+            Self::Permits { last_acquire, .. } => Some(last_acquire),
+        }
+    }
+
+    fn acquire(&mut self, mut num_permits: usize, acquire_clock: VectorClock) -> Result<VectorClock, TryAcquireError> {
+        // Acquiring zero permits is always possible, and is not causally
+        // dependent on any event.
+        if num_permits == 0 {
+            return Ok(VectorClock::new());
+        }
+
+        match self {
+            Self::Infinity(clock) => Ok(clock.clone()),
+            Self::Permits {
+                num_available,
+                permit_clocks,
+                last_acquire,
+            } => {
+                if num_permits <= *num_available {
+                    last_acquire.update(&acquire_clock);
+
+                    // If infinite permits were released, don't decrement.
+                    if *num_available < usize::MAX {
+                        *num_available -= num_permits;
+                    }
+
+                    // Acquire `num_permits` from the available batches. This
+                    // may consume one or more batches from the queue. The
+                    // resulting clock should be the join of all the batches
+                    // used (fully or partially), since the acquiry causally
+                    // depends on the releases that created those batches.
+                    let mut clock = VectorClock::new();
+                    while let Some((batch_size, batch_clock)) = permit_clocks.front_mut() {
+                        clock.update(batch_clock);
+
+                        // An infinite number of permits was released at some
+                        // point, switch variant to reduce future operations.
+                        if *batch_size == usize::MAX {
```

Comment:
We cannot do all of this on `release` because we don't want to completely forget the clocks of the permits that are already in the `permit_clocks`, which (in global time) are earlier events than the release of the "infinity" batch.